### PR TITLE
Move ChunkingCookieManager to a separate Sources project

### PR DESCRIPTION
**Pull request by kichalla**
@Tratcher @rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationOptions.cs [hunk 1]
```diff
@@ -2,7 +2,6 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.ComponentModel;
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationOptions.cs]: will revert the changes in this file

**Comment by davidfowl** [comment]
Wait, you plan to use this allocatey code elsewhere? Where I must know 😄

**Comment by kichalla** [comment]
@davidfowl  in CookieTempDataProvider to chunk potentially large cookies as typically this happens:
cookie value = Base64Encode(DataProtect(BinaryJson(actual-data)))

**Comment by Tratcher** [comment]
Have you tried plugging the new sources package into MVC yet?

**Comment by kichalla** [comment]
@Tratcher yes and it builds and tests run fine.

**Comment by davidfowl** [comment]
> @davidfowl in CookieTempDataProvider to chunk potentially large cookies as this happens:
cookie value = Base64Encode(DataProtect(BinaryJson(actual-data)))

That is terrible. That looks hella inefficient. Where is that code.

**Comment by kichalla** [comment]
In PR currently, https://github.com/aspnet/Mvc/pull/5257/files#diff-8edc4d422c5102070cbdb6f993605c20R75

File: Security.sln [hunk 2]
```diff
@@ -57,6 +57,8 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Owin.Security.Int
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "OpenIdConnect.AzureAdSample", "samples\OpenIdConnect.AzureAdSample\OpenIdConnect.AzureAdSample.xproj", "{3A7AD414-EBDE-4F92-B307-4E8F19B6117E}"
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.Cookies.ChunkingCookieManager.Sources", "src\Microsoft.AspNetCore.Cookies.ChunkingCookieManager.Sources\Microsoft.AspNetCore.Cookies.ChunkingCookieManager.Sources.xproj", "{B6297507-EECF-4114-9B38-6CCE7F9C3EB5}"
```
Comment (rynowak) [inline Security.sln]: I feel like we should drop 'cookies' from the name. `Microsoft.AspNetCore.Cookies` isn't a namespace we ship anywhere

File: src/Microsoft.AspNetCore.Cookies.ChunkingCookieManager.Sources/ChunkingCookieManager.cs [hunk 3]
```diff
@@ -17,6 +20,16 @@ namespace Microsoft.AspNetCore.Authentication.Cookies
     /// </summary>
     public class ChunkingCookieManager : ICookieManager
     {
+#else
+namespace Microsoft.AspNetCore.Cookies
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Cookies.ChunkingCookieManager.Sources/ChunkingCookieManager.cs]: `Microsoft.AspNetCore.Internal`

File: test/Microsoft.AspNetCore.Authentication.Test/Cookies/Infrastructure/CookieChunkingTests.cs [hunk 4]
```diff
@@ -5,8 +5,13 @@
 using Microsoft.AspNetCore.Http;
 using Xunit;
 
-namespace Microsoft.AspNetCore.Authentication.Cookies.Infrastructure
+#if SECURITY
+namespace Microsoft.AspNetCore.Authentication.Cookies
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Authentication.Test/Cookies/Infrastructure/CookieChunkingTests.cs]: This test should be in its own project is possible.

File: Security.sln [hunk 5]
```diff
@@ -352,5 +366,6 @@ Global
 		{A7922DD8-09F1-43E4-938B-CC523EA08898} = {4D2B6A51-2F9F-44F5-8131-EA5CAC053652}
 		{A2B5DC39-68D5-4145-A8CC-6AEAB7D33A24} = {7BF11F3A-60B6-4796-B504-579C67FFBA34}
 		{3A7AD414-EBDE-4F92-B307-4E8F19B6117E} = {F8C0AA27-F3FB-4286-8E4C-47EF86B539FF}
+		{B6297507-EECF-4114-9B38-6CCE7F9C3EB5} = {4D2B6A51-2F9F-44F5-8131-EA5CAC053652}
```
Comment (rynowak) [inline Security.sln]: Make sure to set the solution configuration for the new project to not build it.

File: src/Microsoft.AspNetCore.Authentication.Cookies/Resources.resx [hunk 6]
```diff
@@ -1,126 +0,0 @@
...
-    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
-  </resheader>
-  <resheader name="writer">
-    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
-  </resheader>
-  <data name="Exception_CookieLimitTooSmall" xml:space="preserve">
-    <value>The cookie key and options are larger than ChunksSize, leaving no room for data.</value>
-  </data>
-  <data name="Exception_ImcompleteChunkedCookie" xml:space="preserve">
-    <value>The chunked cookie is incomplete. Only {0} of the expected {1} chunks were found, totaling {2} characters. A client size limit may have been exceeded.</value>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Authentication.Cookies/Resources.resx]: Did this file get deleted or moved?

**Comment by kichalla** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Authentication.Cookies/Resources.resx (hunk 6 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Cookies/Resources.resx]: The only resources were for cooking chunking, and we don't have a way to LOC resources in shared sources.

File: src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ChunkingCookieManager.cs [hunk 7]
```diff
@@ -247,10 +265,10 @@ public void DeleteCookie(HttpContext context, string key, CookieOptions options)
             }
 
             var responseHeaders = context.Response.Headers;
-            var existingValues = responseHeaders[Constants.Headers.SetCookie];
+            var existingValues = responseHeaders["Set-Cookie"];
             if (!StringValues.IsNullOrEmpty(existingValues))
             {
-                responseHeaders[Constants.Headers.SetCookie] = existingValues.Where(value => !rejectPredicate(value)).ToArray();
+                responseHeaders["Set-Cookie"] = existingValues.Where(value => !rejectPredicate(value)).ToArray();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ChunkingCookieManager.cs]: Use the constant from Microsoft.Net.Http.Headers/HeaderNames instead

**Comment by Tratcher** [comment]
One other consideration. Because this writes directly to the headers it will bypass the CookiePolicy middleware (https://github.com/aspnet/Security/tree/dev/src/Microsoft.AspNetCore.CookiePolicy)

File: src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ICookieManager.cs [hunk 8]
```diff
@@ -11,6 +14,12 @@ namespace Microsoft.AspNetCore.Authentication.Cookies
     /// </summary>
     public interface ICookieManager
     {
+#else
+namespace Microsoft.AspNetCore.Internal
+{
+    internal interface ICookieManager
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ICookieManager.cs]: Do you even need the interface for your internal implementation?

File: test/Microsoft.AspNetCore.ChunkingCookieManager.Sources.Test/CookieChunkingTests.cs [hunk 9]
```diff
@@ -5,8 +5,13 @@
 using Microsoft.AspNetCore.Http;
 using Xunit;
 
-namespace Microsoft.AspNetCore.Authentication.Cookies.Infrastructure
+#if SECURITY
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.ChunkingCookieManager.Sources.Test/CookieChunkingTests.cs]: #if SECURITY won't ever be true for this test project, correct?

File: test/Microsoft.AspNetCore.Authentication.Test/project.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
   "buildOptions": {
-    "warningsAsErrors": true
+    "warningsAsErrors": true,
+    "define": ["SECURITY"]
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/project.json]: Why is this needed? Shouldn't the one in Cookies be adequate?

File: src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ICookieManager.cs (hunk 8 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.ChunkingCookieManager.Sources/ICookieManager.cs]: Agreed. Removed it in the latest iteration.

**Comment by kichalla** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.ChunkingCookieManager.Sources.Test/CookieChunkingTests.cs (hunk 9 above)
Comment (Tratcher) [inline test/Microsoft.AspNetCore.ChunkingCookieManager.Sources.Test/CookieChunkingTests.cs]: You don't need an if here, this file doesn't move, you will only ever be testing one version of the src file.

**Comment by kichalla** [comment]
🆙 📅
